File: src/java/main/src/main/java/com/tencent/wcdb/orm/TableBinding.java
// Created by qiuwenchen on 2023/3/30.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.wcdb.orm;

import com.tencent.wcdb.core.PreparedStatement;

public interface TableBinding<T> {
    Class<T> bindingType();

    Field<T>[] allBindingFields();

    Binding baseBinding();

    T extractObject(Field<T>[] fields, PreparedStatement preparedStatement);

    void bindField(T object, Field<T> field, int index, PreparedStatement preparedStatement);

    boolean isAutoIncrement(T object);

    void setLastInsertRowId(T object, long lastInsertRowId);
}

File: src/rust/wcdb_core/src/orm/table_binding.rs
use crate::core::prepared_statement::PreparedStatement;
use crate::orm::binding::Binding;
use crate::orm::field::Field;
use std::any::TypeId;

pub trait TableBinding<T> {
    fn binding_type(&self) -> TypeId;

    fn all_binding_fields(&self) -> Vec<&Field<T>>;

    fn base_binding(&self) -> &Binding;

    fn extract_object(
        &self,
        fields: Vec<Field<T>>,
        prepared_statement: &PreparedStatement,
    ) -> T;

    fn bind_field(
        &self,
        object: &T,
        field: &Field<T>,
        index: usize,
        prepared_statement: &mut PreparedStatement,
    );

    fn is_auto_increment(&self, object: &T) -> bool;

    fn set_last_insert_row_id(&self, object: &mut T, last_insert_row_id: i64);
}
